core/controllers/base.py

```diff
@@ -571,6 +572,21 @@ def _render_exception_json_or_html(self, return_type, values):
                     'error-iframed.mainpage.html', iframe_restriction=None)
             elif values['status_code'] == 503:
                 self.render_template('maintenance-page.mainpage.html')
+            elif values['status_code'] == 404:
+                # Routes for pages are handled at the frontend using angular
+                # router. As they are not registered at the backend,
+                # the backend tries to return error page with 404 status code.
```

Comment:
What if the page is a legitimate 404 -- what happens in the flow where the user goes to /abc (say)?

I'm a bit unsure about the difference between (a) route is not handled by Angular frontend but still exists in general, (b) route is invalid.

Reply:
**All the routes not registered at the backend are handled further by the angular router.**

![proposed_routing_system](https://user-images.githubusercontent.com/24855641/126205645-eb91c013-53e8-4278-85c7-a2ad1baa79c8.png)

We cannot decide that a page is a legitimate 404 at the backend as it may be a case that route is not registered at the backend but is registered with angular router. So, for legitimate 404 the flow with direct to angular router, then angular check it with its own list of routes, and if it does not match it will lazy load the error module (page) and show it to the user.

I realize my mistake that I should have named the variable registered with angular router instead, as all routes which are not found on backend are redirected to angular router (so, in a sense it handle those too).

What the above code does is handling status code. For the legitimate 404 page, the status code will be set 404 and for the valid pages registered with angular router, it will be 200.

All the routes registered on the backend (handlers and pages not yet migrated to router) are not handled by the angular router, (a) case in your question.
(b) A invalid route is further redirected to angular router by rendering `oppia-root.mainpage.html` (unique entry point) and then the angular router will decide whether it is a valid page or a legitimate 404 route.

https://user-images.githubusercontent.com/24855641/126206315-86620979-647f-4c8c-a779-3568fec17560.mp4

![image (2)](https://user-images.githubusercontent.com/24855641/126207985-c048cb26-eb6c-434d-b65c-9337a56264d9.png)
With the above code, status code for the routes in red box is 200 and for those matching green box (legitimate 404) is 404.

Reply:
Thanks, @ashutoshc8101, I think I get it now. But conceptually, this is quite a weird way of handling it. We shouldn't be setting it as a 404 and then changing it to 200.

Instead, for such routes, it would be better to have them all result in 200 at the outset. Could we do something like, add those routes to main.py in a for loop and have them all redirect to the same handler which just serves an "empty" 200 instead?

Reply:
(Feel free to ping or email me if you'd like to discuss the approach further.)

Reply:
I like the new approach and it will make the changes more understandable.

Reply:
made the changes